Add "RETL Scripts For Importing Salesforce Objects Into Unified Profiles" documentation (#6166)
segmentio/segment-docs: src/unified-profiles/create-sql-traits.md

---

> title: RETL Scripts For Importing Salesforce Objects Into Unified Profiles in Flex

Author: ```suggestion
title: RETL Queries for Importing Salesforce Objects Into Unified Profiles in Flex
```

hidden: true
---

> Unified Profiles users can convert Salesforce objects with US phone number patterns (typically (555) 231-7654) into Segment Unify profiles with E.164 phone number formats (+15552317654) to support Unified Profiles lookup.

Reviewer: Third person? or should it be in first person?

> Unified Profiles users can convert Salesforce objects with US phone number patterns (typically (555) 231-7654) into Segment Unify profiles with E.164 phone number formats (+15552317654) to support Unified Profiles lookup.

Author: ```suggestion
You can use the following queries to convert Salesforce objects with US phone number patterns (typically (555) 231-7654) into Segment Unify profiles with E.164 phone number formats (+15552317654). When you update the number patterns, you support Unified Profiles lookup.
```

> Unified Profiles in Flex users can convert Salesforce objects with US phone number patterns (typically (555) 231-7654) into Segment Unify profiles with E.164 phone number formats (+15552317654) to support Unified Profiles lookup.

Author: ```suggestion
You can use the following SQL queries to convert Salesforce objects with US phone number patterns (typically (555) 231-7654) into Segment Unify profiles with E.164 phone number formats (+15552317654), the format required by Unified Profiles in Flex.
```

> Segment created three sample queries for Unified Profiles users to import common Salesforce objects: 

Reviewer: We're giving them scripts to import phone numbers for these types of objects? If so, then:

```suggestion
Segment created three sample queries for Unified Profiles users to import phone numbers from common Salesforce objects: 
```

Author: They're using these scripts as a RETL mapping to both import objects from Salesforce and update the phone number format in one pass, but the more important point here is that they're importing objects

- [Accounts](#accounts)
- [Contacts](#contacts)
- [Leads](#leads)

To selectively import columns, replace the `a*`, `c*`, or `l*` with a list of fields to selectively import. For example, if you wanted to only import the ID, NAME, and ADDRESS fields for your Accounts, you'd replace `a*` with  `a.ID, a.NAME, a.ADDRESS`.

> success ""
> Segment creates a default database table (`segment_flex_unify`) during the [Segment for Flex](/docs/unified-profiles/segment-for-flex/){:target="_blank”} setup process.  

## Accounts

To import Salesforce Accounts into Unified Profiles as Segment Unify profiles, create a RETL mapping with the following format.

> Replace `<database_name.account_table>` with your database name and account table, `PHONE` with the name of the column in your warehouse that contains phone numbers, and `BILLING_COUNTRY` with the name of the column in your warehouse that contains the account's country.  In this example query, the `PHONE` column is the primary contact phone number and what you import as the Profile’s 'phone' trait.

Author: ```suggestion
Replace `<database_name.account_table>` with your database name and account table, `PHONE` with the name of the column in your warehouse that contains phone numbers, and `BILLING_COUNTRY` with the name of the column in your warehouse that contains the account's country.  In this sample query, the `PHONE` column is the primary contact phone number and what you import as the Profile’s 'phone' trait.
```

> The other phone fields for Salesforce Accounts are PersonHomePhone, PersonMobilePhone, & PersonOtherPhone, and could be substituted for `PHONE` in this query. **You can only import one phone number field as the identifier used for lookups in Flex.** 

Author: ```suggestion
The other phone fields for Salesforce Accounts are PersonHomePhone, PersonMobilePhone, & PersonOtherPhone, and could be substituted for `PHONE` in this query. **You can only import one phone number field as the identifier used for lookups in Unified Profiles.** 
```

> The other phone fields for Salesforce Accounts are PersonHomePhone, PersonMobilePhone, & PersonOtherPhone, and could be substituted for `PHONE` in this query. **You can only import one phone number field as the identifier used for lookups in Flex.** 

Author: ```suggestion
The other phone fields for Salesforce Accounts are PersonHomePhone, PersonMobilePhone, & PersonOtherPhone, and could be substituted for `PHONE` in this query. **You can only import one phone number field as the identifier used for lookups in Unified Profiles.** 
```

SELECT 
    a.*,
    CASE 
        WHEN a.PHONE REGEXP '^(\\([0-9]{3}\\) [0-9]{3}-[0-9]{4})$' AND a.BILLING_COUNTRY = 'US'
        THEN CONCAT('+1', REGEXP_REPLACE(a.PHONE, '[^0-9]',''))
        WHEN a.BILLING_COUNTRY != 'US'
        THEN REGEXP_REPLACE(a.PHONE, '[^0-9]','') 
        ELSE a.PHONE
    END as phone,
FROM 
    <database_name.account_table> a
WHERE 
    a.PHONE IS NOT NULL
    AND a.BILLING_COUNTRY IS NOT NULL;
```

> Salesforce objects have several phone number-related fields. In this example query, the “PHONE” column is the primary contact phone number and what you import as the Profile’s 'phone' trait. The other phone fields for Salesforce Accounts are PersonHomePhone, PersonMobilePhone, & PersonOtherPhone, and could be substituted for "PHONE" in this query.

Reviewer: Can you only import one of these phone number types at a time? Or, one per script? Might be worth noting.

Reviewer 2: You can use the same regex to import as many as you want with the proper formatting.

Author: Hi @cherylj - how would you do this? by adding additional WHEN/THEN clauses:
```sql
WHEN a.PHONE REGEXP '^(\\([0-9]{3}\\) [0-9]{3}-[0-9]{4})$' AND a.BILLING_COUNTRY = 'US'
THEN CONCAT('+1', REGEXP_REPLACE(a.PHONE, '[^0-9]',''))
WHEN a.PersonHomePhone REGEXP '^(\\([0-9]{3}\\) [0-9]{3}-[0-9]{4})$' AND a.BILLING_COUNTRY = 'US'
THEN CONCAT('+1', REGEXP_REPLACE(a.PersonHomePhone, '[^0-9]',''))
```
or: 
`WHEN a.PHONE OR WHEN a.PersonHomePhone OR WHEN a.PersonMobilePhone`?

Author: clarified with Cheryl over a Slack huddle - you can only import and format one phone field per account/contact/lead - that phone field would then serve as the identifier that you can lookup in Flex.

> Salesforce objects have several phone number-related fields. In this example query, the “PHONE” column is the primary contact phone number and what you import as the Profile’s 'phone' trait. The other phone fields for Salesforce Accounts are PersonHomePhone, PersonMobilePhone, & PersonOtherPhone, and could be substituted for "PHONE" in this query.

Author: ```suggestion
Salesforce objects have several phone number-related fields. In this example query, the “PHONE” column is the primary contact phone number and what you import as the Profile’s 'phone' trait. The other phone fields for Salesforce Accounts are PersonHomePhone, PersonMobilePhone, & PersonOtherPhone, and could be substituted for "PHONE" in this query. You can import multiple phone fields at a time using this query by including additional WHEN clauses that use the same REGEXP format, replacing a.PHONE with a.<phoneField>. 
```

> ## Accounts
> 
> To import Salesforce Accounts into Unified Profiles as Segment Unify profiles, create a RETL mapping with the following format. 
> 
> Replace `<database_name.account_table>` with your database name and account table, `PHONE` with the name of the column in your warehouse that contains phone numbers, and `BILLING_COUNTRY` with the name of the column in your warehouse that contains the account's country. 
> 
> ``` sql
> SELECT 
>     a.*,
>     CASE 
>         WHEN a.PHONE REGEXP '^(\\([0-9]{3}\\) [0-9]{3}-[0-9]{4})$' AND a.BILLING_COUNTRY = 'US'
>         THEN CONCAT('+1', REGEXP_REPLACE(a.PHONE, '[^0-9]',''))
>         WHEN a.BILLING_COUNTRY != 'US'
>         THEN REGEXP_REPLACE(a.PHONE, '[^0-9]','') 
>         ELSE a.PHONE
>     END as phone,
> FROM 
>     <database_name.account_table> a
> WHERE 
>     a.PHONE IS NOT NULL
>     AND a.BILLING_COUNTRY IS NOT NULL;
> ```
> 
> Salesforce objects have several phone number-related fields. In this example query, the “PHONE” column is the primary contact phone number and what you import as the Profile’s 'phone' trait. The other phone fields for Salesforce Accounts are PersonHomePhone, PersonMobilePhone, & PersonOtherPhone, and could be substituted for "PHONE" in this query.
> 
> After running this query, ‘phone’ can be used as an identifier used for lookups in Unified Profiles.  

Author: ```suggestion
## Accounts

To import Salesforce Accounts into Unified Profiles as Segment Unify profiles, create a RETL mapping with the following format. 

Replace `<database_name.account_table>` with your database name and account table, `PHONE` with the name of the column in your warehouse that contains phone numbers, and `BILLING_COUNTRY` with the name of the column in your warehouse that contains the account's country.  In this example query, the “PHONE” column is the primary contact phone number and what you import as the Profile’s 'phone' trait. 

The other phone fields for Salesforce Accounts are PersonHomePhone, PersonMobilePhone, & PersonOtherPhone, and could be substituted for "PHONE" in this query. **You can only import one phone number field as the identifier used for lookups in Flex.** 

``` sql
SELECT 
    a.*,
    CASE 
        WHEN a.PHONE REGEXP '^(\\([0-9]{3}\\) [0-9]{3}-[0-9]{4})$' AND a.BILLING_COUNTRY = 'US'
        THEN CONCAT('+1', REGEXP_REPLACE(a.PHONE, '[^0-9]',''))
        WHEN a.BILLING_COUNTRY != 'US'
        THEN REGEXP_REPLACE(a.PHONE, '[^0-9]','') 
        ELSE a.PHONE
    END as phone,
FROM 
    <database_name.account_table> a
WHERE 
    a.PHONE IS NOT NULL
    AND a.BILLING_COUNTRY IS NOT NULL;
```

After running this query, ‘phone’ can be used as an identifier used for lookups in Unified Profiles.  
```

## Contacts

To import Salesforce Contacts into Unified Profiles as Segment Unify profiles, create a RETL mapping with the following format. 

Replace `<database_name.contact_table>` with your database name and account table, `PHONE` with the name of the column in your warehouse that contains phone numbers, and `BILLING_COUNTRY` with the name of the column in your warehouse that contains the contact's country. 

> Salesforce objects have several phone number-related fields. In this example query, the `PHONE` column is the primary contact phone number and what you import as the Profile’s 'phone' trait. The other phone fields for Salesforce Contacts are HomePhone, MobilePhone, & OtherPhone, and could be substituted for `PHONE` in this query. **You can only import one phone number field as the identifier used for lookups in Flex.** 

Author: ```suggestion
Salesforce objects have several phone number-related fields. In this sample query, the `PHONE` column is the primary contact phone number and what you import as the Profile’s 'phone' trait. 

The other phone fields for Salesforce Contacts are HomePhone, MobilePhone, & OtherPhone, and could be substituted for `PHONE` in this query. **You can only import one phone number field as the identifier used for lookups in Unified Profiles.** 
```

SELECT 
    c.*,
    CASE 
        WHEN c.PHONE REGEXP '^(\\([0-9]{3}\\) [0-9]{3}-[0-9]{4})$' AND c.BILLING_COUNTRY = 'US'
        THEN CONCAT('+1', REGEXP_REPLACE(c.PHONE, '[^0-9]',''))
        WHEN c.BILLING_COUNTRY != 'US'
        THEN REGEXP_REPLACE(c.PHONE, '[^0-9]','') 
        ELSE c.PHONE
    END as phone,
FROM 
    <database_name.contact_table> c
WHERE 
    c.PHONE IS NOT NULL
    AND c.BILLING_COUNTRY IS NOT NULL;
```

> Salesforce objects have several phone number-related fields. In this example query, the “PHONE” column is the primary contact phone number and what you import as the Profile’s 'phone' trait. The other phone fields for Salesforce Contacts are HomePhone, MobilePhone, & OtherPhone, and could be substituted for "PHONE" in this query.

Reviewer: same as above if we need to clarify how many.

After running this query, you can use ‘phone’ for lookups in Unified Profiles.    

## Leads

To import Salesforce Leads into Unified Profiles as Segment Unify profiles, create a RETL mapping with the following format. 

Replace `<database_name.lead_table>` with your database name and lead table, `PHONE` with the name of the column in your warehouse that contains phone numbers, and `BILLING_COUNTRY` with the name of the column in your warehouse that contains the lead's country. 

> Salesforce objects have several phone number-related fields. In this example query, the `PHONE` column is the primary contact phone number and what you import as the Profile’s 'phone' trait. The other phone fields for Salesforce Leads are HomePhone, MobilePhone, & OtherPhone, and could be substituted for `PHONE` in this query. **You can only import one phone number field as the identifier used for lookups in Flex.** 

Author: ```suggestion
Salesforce objects have several phone number-related fields. In this sample query, the `PHONE` column is the primary contact phone number and what you import as the Profile’s 'phone' trait. 

The other phone fields for Salesforce Leads are HomePhone, MobilePhone, & OtherPhone, and could be substituted for `PHONE` in this query. **You can only import one phone number field as the identifier used for lookups in Unified Profiles.** 
```

---
title: RETL Scripts For Importing Salesforce Objects Into Unified Profiles
hidden: true
---
Unified Profiles users can convert Salesforce objects with US phone number patterns (typically (555) 231-7654) into Segment Unify profiles with E.164 phone number formats (+15552317654) to support Unified Profiles lookup.

> Segment created three sample queries for Unified Profiles users to import common Salesforce objects: 

Reviewer 2: It might be worth calling out that database table created by the Flex Unify custom onboarding (so, not pre-existing Segment customers) is `segment_flex_unify` somewhere here in this doc.

- [Accounts](#accounts)
- [Contacts](#contacts)
- [Leads](#leads)

> To create a query that selects a different field, replace the `a`, `c`, or `l` with a list of fields to selectively import. 

Reviewer 2: This isn't quite right.  The `a`, `c`, or `l` are shorthand references the tables not fields.

To only select a subset of columns, you'd change this `a.*`, to something like `a.ID, a.NAME, a.ADDRESS`, etc.

> To create a query that selects a different field, replace the `a`, `c`, or `l` with a list of fields to selectively import. 

Author: ```suggestion
To selectively import columns, replace the `a*`, `c*`, or `l*` with a list of fields to selectively import. 
```

## Accounts

To import Salesforce Accounts into Unified Profiles as Segment Unify profiles, create a RETL mapping with the following format. 

> Replace `<database_name.account_table>` with your database name and account table, `PHONE` with the name of the column in your warehouse that contains phone numbers, and `BILLING_COUNTRY` with the name of the column in your warehouse that contains the account's billing country. 

Reviewer 2: Looks like this line also exists for contacts and leads.

> Replace `<database_name.account_table>` with your database name and account table, `PHONE` with the name of the column in your warehouse that contains phone numbers, and `BILLING_COUNTRY` with the name of the column in your warehouse that contains the account's billing country. 

Author: ```suggestion
Replace `<database_name.account_table>` with your database name and account table, `PHONE` with the name of the column in your warehouse that contains phone numbers, and `BILLING_COUNTRY` with the name of the column in your warehouse that contains the account's country. 
```

``` sql
SELECT 
    a.*,
    CASE 
        WHEN a.PHONE REGEXP '^(\\([0-9]{3}\\) [0-9]{3}-[0-9]{4})$' AND a.BILLING_COUNTRY = 'US'
        THEN CONCAT('+1', REGEXP_REPLACE(a.PHONE, '[^0-9]',''))
        WHEN a.BILLING_COUNTRY != 'US'
        THEN REGEXP_REPLACE(a.PHONE, '[^0-9]','') 
        ELSE a.PHONE
    END as phone,
FROM 
    <database_name.account_table> a
WHERE 
    a.PHONE IS NOT NULL
    AND a.BILLING_COUNTRY IS NOT NULL;
```

Salesforce objects have several phone number-related fields. In this example query, the “PHONE” column is the primary contact phone number and what you import as the Profile’s 'phone' trait. The other phone fields for Salesforce Accounts are PersonHomePhone, PersonMobilePhone, & PersonOtherPhone, and could be substituted for "PHONE" in this query.

> After running this query, 'phone' would be an Identity used for lookups in Unified Profiles.    

Reviewer 2: I'd change this line to:
After running this query, ‘phone’ can be used as an identifier used for lookups in Unified Profiles.

Reviewer 2: Looks like this line also exists for contacts and leads.

> After running this query, 'phone' would be an Identity used for lookups in Unified Profiles.    

Author: ```suggestion
After running this query, ‘phone’ can be used as an identifier used for lookups in Unified Profiles.  
```

## Contacts

To import Salesforce Contacts into Unified Profiles as Segment Unify profiles, create a RETL mapping with the following format. 

> Replace `<database_name.contact_table>` with your database name and account table, `PHONE` with the name of the column in your warehouse that contains phone numbers, and `BILLING_COUNTRY` with the name of the column in your warehouse that contains the contact's billing country.

Author: ```suggestion
Replace `<database_name.contact_table>` with your database name and account table, `PHONE` with the name of the column in your warehouse that contains phone numbers, and `BILLING_COUNTRY` with the name of the column in your warehouse that contains the contact's country.
```

``` sql
SELECT 
    c.*,
    CASE 
        WHEN c.PHONE REGEXP '^(\\([0-9]{3}\\) [0-9]{3}-[0-9]{4})$' AND c.BILLING_COUNTRY = 'US'
        THEN CONCAT('+1', REGEXP_REPLACE(c.PHONE, '[^0-9]',''))
        WHEN c.BILLING_COUNTRY != 'US'
        THEN REGEXP_REPLACE(c.PHONE, '[^0-9]','') 
        ELSE c.PHONE
    END as phone,
FROM 
    <database_name.contact_table> c
WHERE 
    c.PHONE IS NOT NULL
    AND c.BILLING_COUNTRY IS NOT NULL;
```

Salesforce objects have several phone number-related fields. In this example query, the “PHONE” column is the primary contact phone number and what you import as the Profile’s 'phone' trait. The other phone fields for Salesforce Contacts are HomePhone, MobilePhone, & OtherPhone, and could be substituted for "PHONE" in this query.

> After running this query, 'phone' would be an Identity used for lookups in Unified Profiles.    

Author: ```suggestion
After running this query, ‘phone’ can be used as an identifier used for lookups in Unified Profiles.   
```

## Leads

To import Salesforce Leads into Unified Profiles as Segment Unify profiles, create a RETL mapping with the following format. 

Replace `<database_name.lead_table>` with your database name and lead table, `PHONE` with the name of the column in your warehouse that contains phone numbers, and `BILLING_COUNTRY` with the name of the column in your warehouse that contains the lead's billing country. 

``` sql
SELECT 
    l.*,
    CASE 
        WHEN l.PHONE REGEXP '^(\\([0-9]{3}\\) [0-9]{3}-[0-9]{4})$' AND l.BILLING_COUNTRY = 'US'
        THEN CONCAT('+1', REGEXP_REPLACE(l.PHONE, '[^0-9]',''))
        WHEN l.BILLING_COUNTRY != 'US'
        THEN REGEXP_REPLACE(l.PHONE, '[^0-9]','') 
        ELSE l.PHONE
    END as phone,
FROM 
    <database_name.lead_table> l
WHERE 
    l.PHONE IS NOT NULL
    AND l.BILLING_COUNTRY IS NOT NULL;
```

Salesforce objects have several phone number-related fields. In this example query, the “PHONE” column is the primary contact phone number and what you import as the Profile’s 'phone' trait. The other phone fields for Salesforce Leads are HomePhone, MobilePhone, & OtherPhone, and could be substituted for "PHONE" in this query.

> After running this query, 'phone' would be an Identity used for lookups in Unified Profiles.  

Author: ```suggestion
After running this query, ‘phone’ can be used as an identifier used for lookups in Unified Profiles.
```
